rust_language/resolve.go

```diff
@@ -148,8 +148,57 @@ func maybeSetAttrStrings(r *rule.Rule, attr string, val []string) {
 	}
 }
 
+func (l *rustLang) resolveCrateVersion(cfg *rustConfig, c *config.Config,
+	crateDeps map[string]CargoDependency, lockfileCrate LockfileCrate,
+	from label.Label) string {
+	var versionRequirement string
+	if cargoDep, ok := crateDeps[lockfileCrate.Name]; ok {
+		// This dependency is specified in the Cargo.toml file of the
+		// workspace member.
+		if cargoDep.workspace || cargoDep.version == "" {
+			// The version requirement is inherited from the workspace
+			// manifest, therefore, take it from there.
+			versionRequirement, ok = cfg.WorkspaceManifest.Dependencies[lockfileCrate.Name]
+			if !ok {
+				l.Log(c, logFatal, from, "multiversion crate %s is inherited from the workspace manifest, but it was not declared there", lockfileCrate.Name)
+			}
+		} else {
+			// Otherwise the version requirement must be directly specified
+			// in the Cargo.toml file of the workspace member.
+			versionRequirement = cargoDep.version
+		}
+	} else {
+		// The dependency is not specified in the Cargo.toml file of the
+		// workspace member. It probably should have been, but it is logical
+		// to default to the workspace version with good chances for
+		// success.
+		versionRequirement, ok = cfg.WorkspaceManifest.Dependencies[lockfileCrate.Name]
+		if !ok {
+			l.Log(c, logFatal, from, "multiversion crate %s is used but declared neither in crate manifest nor in workspace manifest", lockfileCrate.Name)
+		}
+	}
+	if versionRequirement == "" {
+		l.Log(c, logFatal, from, "crate %s is declared as a multiversion crate, but no version spec could be deduced", lockfileCrate.Name)
+	}
+	version := ""
+	// We have obtained the versionRequirement, now match it against the
+	// available complete versions from the Cargo lock file:
+	for _, candidateVersion := range lockfileCrate.Versions {
+		if strings.HasPrefix(candidateVersion, versionRequirement) {
+			version = candidateVersion
+			break
+		}
+	}
+	if version == "" {
+		// We haven't found anything, that's not normal, abort.
+		l.Log(c, logFatal, from, "could not resolve the version for %s", lockfileCrate.Name)
```

Comment:
Arguably this should be a warning like how other dependency resolution failures are warnings.